pkg/migration/migration.go

```diff
@@ -0,0 +1,586 @@
+	return &Handler{
+		client: ctrl.GetKubernetesClient(),
+		driver: ctrl.Driver,
+	}
+}
+
+// Start starts the migration
+func (h *Handler) Start() {
+	ds, err := h.getPortworxDaemonSet()
+	if err != nil {
+		logrus.Errorf("Migration failed to start. %v", err)
+		return
+	} else if ds == nil {
+		logrus.Infof("Migration is not needed")
+		return
+	}
+
+	cluster, err := h.createStorageClusterIfAbsent(ds)
+	if err != nil {
+		logrus.Errorf("Migration failed to create StorageCluster. %v", err)
+		return
+	}
+
+	wait.PollInfinite(migrationRetryInterval, func() (bool, error) {
+		cluster, isNeeded := h.isMigrationNeeded(cluster)
+		if !isNeeded {
+			return true, nil
+		} else if cluster == nil {
+			return false, nil
+		}
+
+		if !h.isMigrationApproved(cluster) {
+			return false, nil
+		}
+
+		if err := h.startMigration(cluster); err != nil {
+			return false, nil
+		}
+		logrus.Infof("Migration completed successfully")
+		return true, nil
+	})
+}
+
+func (h *Handler) createStorageClusterIfAbsent(ds *appsv1.DaemonSet) (*corev1.StorageCluster, error) {
+	clusterName := getPortworxClusterName(ds)
+	stc := &corev1.StorageCluster{}
+	err := h.client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      clusterName,
+			Namespace: ds.Namespace,
+		},
+		stc,
+	)
+	if err == nil {
+		return stc, nil
+	} else if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	stc, err = h.createStorageCluster(ds)
+	// TODO: Create dry run specs
+	return stc, err
+}
+
+func (h *Handler) startMigration(cluster *corev1.StorageCluster) error {
+	// TODO: Implement this
+	// 1. Backup existing specs
+	// 2. Migrate daemonset pods
+	// 3. Migrate components
+	// 4. Delete daemonset and components
+	return nil
+}
+
+func (h *Handler) createStorageCluster(
+	ds *appsv1.DaemonSet,
+) (*corev1.StorageCluster, error) {
+	stc := h.constructStorageCluster(ds)
+
+	// TODO: Handle enable stork
+	// TODO: Handle enable autopilot
+	// TODO: Handle enable monitoring
+
+	err := h.client.Create(context.TODO(), stc)
+	if err == nil {
+		stc.Status.Phase = phaseAwaitingApproval
+		err = h.client.Status().Update(context.TODO(), stc)
+	}
+	return stc, err
+}
+
+func (h *Handler) constructStorageCluster(ds *appsv1.DaemonSet) *corev1.StorageCluster {
+	cluster := &corev1.StorageCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ds.Namespace,
+			Annotations: map[string]string{
+				constants.AnnotationMigrationApproved: "false",
+			},
+		},
+	}
+
+	c := getPortworxContainer(ds)
+	cluster.Spec.Image = c.Image
+	cluster.Spec.ImagePullPolicy = c.ImagePullPolicy
+	if len(ds.Spec.Template.Spec.ImagePullSecrets) > 0 {
+		cluster.Spec.ImagePullSecret = stringPtr(ds.Spec.Template.Spec.ImagePullSecrets[0].Name)
+	}
+
+	envMap := map[string]*v1.EnvVar{}
+	kvdbSecret := struct {
+		ca       string
+		cert     string
+		key      string
+		aclToken string
+		userpwd  string
+	}{}
+	miscArgs := ""
+
+	for i := 0; i < len(c.Args); i++ {
```

Comment:
feel it's easy to mess up this for loop index, probably could come up with some better ideas to parse them in the future?

Reply:
Yes, I agree. Also, the assumption is that the arguments are always valid. Generally if we pass invalid arguments like arg with no value, non-existent arg, or incorrect value type, oci-mon or portworx will fail to come up. 
This means the existing cluster itself is not up.
But I will see if we can come up with something better and handle more scenarios as we mature this.

Reply:
Let me know if you have a better idea.